src/hotspot/cpu/arm/gc/shared/barrierSetAssembler_arm.cpp

```diff
@@ -195,3 +198,43 @@ void BarrierSetAssembler::incr_allocated_bytes(MacroAssembler* masm, RegisterOrC
   // Unborrow the Rthread
   __ sub(Rthread, Ralloc, in_bytes(JavaThread::allocated_bytes_offset()));
 }
+
+void BarrierSetAssembler::nmethod_entry_barrier(MacroAssembler* masm) {
+
+  BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+
+  Register tmp0 = Rtemp;
+  Register tmp1 = R5; // wrong register here will lead to crash
+
+  if (bs_nm == NULL) {
+    return;
+  }
+
+  Label skip, guard;
+  Address thread_disarmed_addr(Rthread, in_bytes(bs_nm->thread_disarmed_offset()));
+
+  __ block_comment("nmethod_barrier begin");
+  __ ldr_label(tmp0, guard);
+
+  // Subsequent loads of oops must occur after load of guard value.
+  // BarrierSetNMethod::disarm sets guard with release semantics.
+  __ membar(MacroAssembler::LoadLoad, tmp1);
```

Comment:
Is my assumption correct that ARM32 supports neither ZGC nor ShenandoahGC?
If so then no memory barrier is needed as no oops are modified concurrently with G1 (and other GCs are not concurrent).
I'd recomment to implement `BarrierSetAssembler::nmethod_patching_type()` as on AARCH64 and replace the membar with an assertion that `NMethodPatchingType::stw_instruction_and_data_patch` is in effect.
@fisk wouldn't it be good if there was a shared method to query if oops (other values?) are modified concurrently?

Reply:
I'm second to the proposal to move nmethod_patching_type() to shared code.